### DISCUSSION TITLE
unittest setup reportUninitializedInstanceVariable=true

### DISCUSSION THREAD

**Discussion by mbyt** [microsoft/pyright#4834] [Q&A] [answered]
I stumbled over the following:

```
> cat pyright_unittest_setup_uninitialized_instance.py
```
```python
# pyright: reportUninitializedInstanceVariable=true
import unittest

class TestDataset(unittest.TestCase):
    def setUp(self):
        self.dataset = [7, 8, 9]

    def test_dataset(self):
        self.assertEqual(self.dataset[2], 9)
```

```
C:\s\scratch via 🐍 v3.11.2 took 2s
❯ : pyright pyright_unittest_setup_uninitialized_instance.py
No configuration file found.
No pyproject.toml file found.
stubPath C:\s\scratch\typings is not a valid directory.
Assuming Python platform Windows
Searching for source files
Found 1 source file
pyright 1.1.300
C:\s\scratch\pyright_unittest_setup_uninitialized_instance.py
  C:\s\scratch\pyright_unittest_setup_uninitialized_instance.py:7:14 - error: Instance variable "dataset" is not initialized in the class body or __init__ method (reportUninitializedInstanceVariable)
1 error, 0 warnings, 0 informations
Completed in 0.561sec
```

In usual python classes, `__init__` is the right place for initialization. For unittests, to my understanding one should use `setup` for test initialization instead of `__init__`.

What would be the pyright way to deal with this?

**Comment by erictraut** [comment, marked as answer]
You can't use `reportUninitializedInstanceVariable` in this case. This check assumes that you're using a specific design pattern whereby all instance variables are initialized in the `__init__` method. So the answer to your question is "don't use `reportUninitializedInstanceVariable` for your unit tests".

**Reply by mbyt** [reply]
alright, thanks for the fast feedback!

**Comment by ms-lolo** [comment]
hey @erictraut, can you help me understand if it's possible to disable this config only in my test directory? i would like to leave it enabled for everything in src/ but i'm not seeing how to control features at the directory level. i would like to avoid adding the ignore at the top of every test.

as an example, `ruff` allows rules to be defined like this:

```
[tool.ruff.lint.per-file-ignores]
# Our tests don't need to document things as much as other public APIs.
"test/python/**/*.py" = [
    "D100", # D100: Missing docstring in public module
    "D101", # D101: Missing docstring in public class
    "D104", # D104: Missing docstring in public package
]
```

**Reply by erictraut** [reply]
Pyright doesn't support per-directory or per-file configurations in the config file. It does support [per-file overrides using comments.](https://microsoft.github.io/pyright/#/comments?id=file-level-type-controls)

**Reply by ms-lolo** [reply]
ah that's unfortunate, but thanks for the quick response

**Reply by efine-vivodyne** [reply]
Is it a deliberate philosophical decision to not support per-directory configurations in the config file?  or is it just something that hasn't been implemented yet?

**Reply by erictraut** [reply]
@efine-vivodyne, pyright now supports different type checking settings per [execution environment](https://microsoft.github.io/pyright/#/configuration?id=execution-environment-options), and each execution environment has a different root directory.

**Reply by efine-vivodyne** [reply]
That's an exciting development.

But, am I missing it when I'm reading the docs?  or does it still not allow different rules to be specified per directory (e.g. ignoring `reportUninitializedInstanceVariable` in `tests` folder)

**Reply by erictraut** [reply]
The documentation says: 

> In addition, any of the [type check diagnostics settings](https://microsoft.github.io/pyright/#/configuration?id=type-check-diagnostics-settings) listed above can be specified. These settings act as overrides for the files in this execution environment.

You can create an execution environment for your `tests` directory and specify that you want `reportUnitializedInstanceVariable` to be `false`.

**Reply by efine-vivodyne** [reply]
ah, great!  I missed that line, and didn't see any examples of that in the Sample Config File.  Thanks!

**Reply by erictraut** [reply]
Yeah, it would be good to have examples in the sample config. I've updated the documentation to add this.
